libvast/src/segment.cpp

```diff
@@ -0,0 +1,201 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include <caf/stream_deserializer.hpp>
+
+#include <caf/detail/scope_guard.hpp>
+
+#include "vast/bitmap_algorithms.hpp"
+#include "vast/const_table_slice_handle.hpp"
+#include "vast/ids.hpp"
+#include "vast/segment.hpp"
+#include "vast/si_literals.hpp"
+#include "vast/table_slice.hpp"
+
+#include "vast/detail/assert.hpp"
+#include "vast/detail/byte_swap.hpp"
+#include "vast/detail/coded_deserializer.hpp"
+#include "vast/detail/coded_serializer.hpp"
+#include "vast/detail/narrow.hpp"
+
+namespace vast {
+
+using namespace binary_byte_literals;
+
+namespace {
+
+template <class T>
+T to_little_endian(T x) {
+  return detail::swap<detail::host_endian, detail::little_endian>(x);
+}
+
+template <class T>
+T from_little_endian(T x) {
+  return detail::swap<detail::little_endian, detail::host_endian>(x);
+}
+
+segment::header make_header(chunk_ptr chunk) {
+  VAST_ASSERT(chunk->size() >= sizeof(segment::header));
+  auto hdr = reinterpret_cast<const segment::header*>(chunk->data());
+  segment::header result;
+  result.magic = from_little_endian(hdr->magic);
+  result.version = from_little_endian(hdr->version);
+  result.id = hdr->id;
+  result.payload_offset = from_little_endian(hdr->payload_offset);
+  return result;
+}
+
+} // namespace <anonymous>
+
+caf::expected<segment_ptr> segment::make(caf::actor_system& sys, chunk_ptr chunk) {
+  VAST_ASSERT(chunk != nullptr);
+  if (chunk->size() < sizeof(header))
+    return make_error(ec::unspecified, "segment too small", chunk->size());
```

Comment:
I'd rather add a new error code instead of using `ec::unspecified` or use something more specific like `invalid_argument` for example.

Reply:
Makes sense.